packages/fast-colors/app/src/PaletteEditor.tsx

```diff
@@ -0,0 +1,199 @@
+/*---------
+  IMPORTS
+----------*/
+
+import React from "react";
+import manageJss, { ComponentStyles } from "@microsoft/fast-jss-manager-react";
+import PaletteEditorPane from "./PaletteEditorPane/PaletteEditorPane";
+import { IPaletteEditorPaneParamsChangedArgs } from "./PaletteEditorPane/PaletteEditorPane";
+import PaletteDisplay from "./PaletteDisplay/PaletteDisplay";
+import { ColorInterpolationSpace, ColorRGBA64, parseColor } from "../../src/colorlib";
+
+/*------------
+  JSS STYLES
+-------------*/
+
+interface IPaletteEditorNameContract {
+  outerContainer: string;
+  mainPane: string;
+  editorPane: string;
+  newPalette: string;
+}
+
+const styles: ComponentStyles<IPaletteEditorNameContract, any> = {
+  outerContainer: {
+    display: "grid",
+    gridTemplateColumns: "auto 300px"
+  },
+  mainPane: {
+    gridColumn: "1 1",
+    display: "flex",
+    flexFlow: "row wrap",
+    justifyContent: "space-evenly",
+    alignItems: "center"
+  },
+  editorPane: {
+    gridColumn: "2 2"
+  },
+  newPalette: {
+    width: "200px",
+    height: "100px",
+    margin: "8px 8px 8px 8px",
+    display: "flex",
+    flexFlow: "row nowrap",
+    alignItems: "center",
+    justifyContent: "center"
+  }
+};
+
+/*----------------
+  PROPS & STATES
+-----------------*/
+
+interface IPaletteEditorProps {
+  managedClasses: IPaletteEditorNameContract;
+}
+
+interface IPaletteEditorState {
+  paletteList: ColorRGBA64[];
+  steps?: number;
+  interpolationMode?: ColorInterpolationSpace;
+  scaleColorLight?: ColorRGBA64;
+  scaleColorDark?: ColorRGBA64;
+  clipLight?: number;
+  clipDark?: number;
+  saturationAdjustmentCutoff?: number;
+  saturationLight?: number;
+  saturationDark?: number;
+  overlayLight?: number;
+  overlayDark?: number;
+  multiplyLight?: number;
+  multiplyDark?: number;
+}
+
+/*------------------
+  CLASS DEFINITION
+-------------------*/
+
+class PaletteEditor extends React.Component<
+  IPaletteEditorProps,
+  IPaletteEditorState
+> {
+  /*-------------
+    CONSTRUCTOR
+  --------------*/
+
+  constructor(props: IPaletteEditorProps) {
+    super(props);
+    this.state = {
+      paletteList: [
+        parseColor("#e74856")!,

```

Comment:
I'm not familiar with what the `!` is doing here - can you explain that? I also see this below when accessing properties of a `ColorRGBA64`

Reply:
In Typescript, the ! operator can be used to assert that an object is not undefined. EG: in this case the ! asserts that the returned value is not undefined.

Reply:
Interesting, I've never seen it used _after_ an object before

Reply:
Super cool - learn something new every day.